### PR TITLE
Send ban parent per outstanding child connection

### PR DESCRIPTION
**Pull request by dnhatn**
Few things that prevent us from supporting cross-clusters tasks cancellation:

1. A child task gets [canceled](https://github.com/elastic/elasticsearch/blob/570f03d3bea63290a0e927b9e302657d01e7022d/server/src/main/java/org/elasticsearch/tasks/TaskManager.java#L407-L417) if a node with the parent task leaves the cluster. This doesn't hold for a cross-clusters task as the parent task does not belong to the remote cluster. This issue was resolved in https://github.com/elastic/elasticsearch/pull/56620, where we cancel child tasks on disconnect instead.

2. Ban action [isn't registered](https://github.com/elastic/elasticsearch/blob/af2f084bff420e8ef367ffa4b3608c1e35b2b800/server/src/main/java/org/elasticsearch/tasks/TaskCancellationService.java#L56-L57) on proxy nodes

3. Banned parent markers are removed when a node with the parent task leaves the cluster. This issue is similar to the first issue.

~This commit tries to address the third issue by periodically (every 5 minutes) sending heartbeats for banned parent markers and removing markers that haven't received heartbeats for some time (30 minutes).~

This commit tries to address the third issue by sending a ban for each outstanding child connection and removing it when all associated connections disconnect.

### CODE REVIEW

**Comment by dnhatn** [comment]
@DaveCTurner I looked into your proposal to remove a ban once all associated child connections are disconnected. However, it isn't watertight as we can remove a ban before an in-flight child task start. Any suggestions are welcomed.

**Comment by DaveCTurner** [comment]
> However, it isn't watertight as we can remove a ban before an in-flight child task start. Any suggestions are welcomed.

Hmm wouldn't that mean the child task is going to respond on a connection that's was closed before the task even started? If so we shouldn't even start the task.

**Comment by dnhatn** [comment]
I should have been more clear. We do not prevent a child task from starting in this scenario:

1. The parent task sends message m1 to start a child task t1 on node-1 using connection c1
2. The parent task also sends message m2 to start a child task t2 on node-1 using connection c2
3. Message m1 arrives, and node-1 starts child task t1
4. The parent task is canceled
5. We send a ban message b1 to node-1 (to cancel or prevent t1/t2 from starting)
6. If the connection c1 is disconnected and we remove the ban b1, then node-1 will start child task t2 when the message m2 arrives

**Comment by DaveCTurner** [comment]
I see. Could we indicate in the ban message that there are already in-flight tasks on connections c1 and c2, so it doesn't get removed while c2 is still open?

**Comment by DaveCTurner** [comment]
On reflection that sounds hard, we have no way to refer to a specific connection that the other end can understand, except of course sending the ban over every relevant connection.

**Comment by dnhatn** [comment]
Another issue is proxy connections. A proxy connection consists of two connections: the source node to the proxy node and the proxy node to the target node. While the former will never change, the latter can change anytime. I think we can't use the connection info to remove ban markers.

**Comment by DaveCTurner** [comment]
Right, that was what I meant earlier about having the proxy also proxy a disconnect event. We can't describe a specific set of connections to a remote cluster, but we can say "the connection on which this message was received".

**Comment by dnhatn** [comment]
@DaveCTurner I implemented a prototype that does not require heartbeats for the ban markers. In the prototype, we send a ban for each outstanding child connection and remove bans when all associated connections of that ban disconnect.

I've updated this PR to send bans per child connection and will open a follow-up to associate connections to bans. This PR is ready for review again. Can you please take another look? Thank you!

**Comment by dnhatn** [comment]
run elasticsearch-ci/1

**Comment by dnhatn** [comment]
I also included (but reverted from this PR) the commit https://github.com/elastic/elasticsearch/pull/65443/commits/143620d5ef278a97e4420d8c7ceca2b78fb9da67 that tracks bans for each channel and removes them when their associated channels are closed.

**Review by DaveCTurner — COMMENTED** [review]
Hmm, I think this still isn't quite watertight, because a `Connection` is a bundle of channels so messages to a single `Connection` are not fully ordered. I think it's possible that we could send a task T on one channel that takes a very long time to arrive, then on a different channel send the ban and receive a response, then clear the ban and receive a response, and only then does task T arrive at the target.

Edit: I see that I caused this confusion: I said "connection" in the earlier discussion. I meant TCP connection, i.e. channel.

**Comment by dnhatn** [comment]
@DaveCTurner Thanks for looking. That's a good point. I think we can make it tighter by sending a ban for every pair of Connection and TransportRequestOptions.Type, which is used to pick a TCP channel. WDYT?

**Comment by DaveCTurner** [comment]
> I think we can make it tighter by sending a ban for every pair of Connection and TransportRequestOptions.Type, which is used to pick a TCP channel. WDYT?

That has the same problem: each type of channel is a bundle of TCP connections. Some bundles have size 1, but they're larger for things where cancellation matters, e.g. searches.

However I'm now thinking that working at the `Connection` level _is_ ok after all. We don't lift the ban until we know that all the child tasks have completed, which means they were either processed by the remote node or else we disconnected. If we disconnected then the remote tasks may still start up but the remote will eventually find out about the disconnection and cancel those tasks too. I thought we were relying on the ordering within a single channel, but I now think it's enough that we tear down the whole `Connection` if any of its individual channels disconnects.

Could you confirm that we're ok if occasionally some remote tasks start up after the ban has been lifted, but only if the `Connection` is destroyed and those tasks will therefore be cancelled eventually. The timeliness of that cancellation depends on things like TCP keepalives, so may take several minutes.

**Comment by dnhatn** [comment]
> Meanwhile the remote might not notice the disconnection for some time and would therefore continue processing any the child tasks.

@DaveCTurner Thank you for iterating on this. The current implementation has a similar issue. I think it's okay because it's not a safety problem but rather wasting resources for a short window.

File: server/src/main/java/org/elasticsearch/transport/TransportService.java [hunk 1]
```diff
@@ -603,7 +607,7 @@ public void removeConnectionListener(TransportConnectionListener listener) {
             final TransportResponseHandler<T> delegate;
             if (request.getParentTask().isSet()) {
                 // TODO: capture the connection instead so that we can cancel child tasks on the remote connections.
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/transport/TransportService.java:609]:
```suggestion
```

TODO is done :tada:

**Review by DaveCTurner — APPROVED** [review]
Ok thanks for confirming, this LGTM then. Neatly done, I like this change 😄

**Comment by dnhatn** [comment]
@DaveCTurner Thanks so much for reviewing and suggesting this direction :)

**Comment by dnhatn** [comment]
@elasticmachine update branch

**Pull request merged**